File: src/applications/likwid-powermeter.lua
#!<INSTALLED_BINPREFIX>/likwid-lua
--[[
 * =======================================================================================
 *
 *      Filename:  likwid-powermeter.lua
 *
 *      Description:  An application to get information about power
 *      consumption on architectures implementing the RAPL interface.
 *
 *      Version:   <VERSION>
 *      Released:  <DATE>
 *
 *      Author:   Thomas Gruber (tr), thomas.roehl@gmail.com
 *      Project:  likwid
 *
 *      Copyright (C) 2016 RRZE, University Erlangen-Nuremberg
 *
 *      This program is free software: you can redistribute it and/or modify it under
 *      the terms of the GNU General Public License as published by the Free Software
 *      Foundation, either version 3 of the License, or (at your option) any later
 *      version.
 *
 *      This program is distributed in the hope that it will be useful, but WITHOUT ANY
 *      WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 *      PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 *      You should have received a copy of the GNU General Public License along with
 *      this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * =======================================================================================
]]
package.path = '<INSTALLED_PREFIX>/share/lua/?.lua;' .. package.path

local likwid = require("likwid")

print_stdout = print
print_stderr = function(...) for k,v in pairs({...}) do io.stderr:write(v .. "\n") end end

local function version()
    print_stdout(string.format("likwid-powermeter -- Version %d.%d.%d (commit: %s)",likwid.version,likwid.release,likwid.minor,likwid.commit))
end

local function examples()
    print_stdout("Examples:")
    print_stdout("Measure the power consumption for 4 seconds on socket 1")
    print_stdout("likwid-powermeter -s 4 -c 1")
    print_stdout("")
    print_stdout("Use it as wrapper for an application to measure the energy for the whole execution")
    print_stdout("likwid-powermeter -c 1 ./a.out")
end

local function usage()
    version()
    print_stdout("A tool to print power and clocking information on x86 CPUs.\n")
    print_stdout("Options:")
    print_stdout("-h, --help\t Help message")
    print_stdout("-v, --version\t Version information")
    print_stdout("-V, --verbose <level>\t Verbose output, 0 (only errors), 1 (info), 2 (details), 3 (developer)")
    print_stdout("-M <0|1>\t\t Set how MSR registers are accessed, 0=direct, 1=accessDaemon")
    print_stdout("-c <list>\t\t Specify sockets to measure")
    print_stdout("-i, --info\t Print information from MSR_PKG_POWER_INFO register and Turbo mode")
    print_stdout("-s <duration>\t Set measure duration in us, ms or s. (default 2s)")
    print_stdout("-p\t\t Print dynamic clocking and CPI values, uses likwid-perfctr")
    print_stdout("-t\t\t Print current temperatures of all hardware threads")
    print_stdout("-f\t\t Print current temperatures in Fahrenheit")
    print_stdout("")
    examples()
end

local config = likwid.getConfiguration();

print_info = false
stethoscope = false
fahrenheit = false
print_temp = false
verbose = 0
if config["daemonMode"] < 0 then
    access_mode = 1
else
    access_mode = config["daemonMode"]
end
time_interval = 2
time_orig = "2s"
read_interval = 10
cpustr = nil
cpuinfo = likwid.getCpuInfo()
--if cpuinfo["isIntel"] == 1 then
--    domainList = {"PKG", "PP0", "PP1", "DRAM", "PLATFORM"}
--else
--    domainList = {"CORE", "PKG"}
--end

for opt,arg in likwid.getopt(arg, {"V:", "c:", "h", "i", "M:", "p", "s:", "v", "f", "t", "help", "info", "version", "verbose:"}) do
    if (type(arg) == "string") then
        local s,e = arg:find("-");
        if s == 1 then
            print_stderr(string.format("Argument %s to option -%s starts with invalid character -.", arg, opt))
            print_stderr("Did you forget an argument to an option?")
            os.exit(1)
        end
    end
    if opt == "h" or opt == "help" then
        usage()
        os.exit(0)
    elseif opt == "v" or opt == "version" then
        version()
        os.exit(0)
    elseif (opt == "c") then
        error("-c is currently unsupported. Please run likwid-powermeter with likwid-pin")
        cpustr = arg
    elseif (opt == "M") then
        access_mode = tonumber(arg)
        if (access_mode == nil) then
            print_stderr("Access mode (-M) must be an number")
            usage()
            os.exit(1)
        elseif (access_mode < 0) or (access_mode > 1) then
            print_stderr(string.format("Access mode (-M) %d not valid.",access_mode))
            usage()
            os.exit(1)
        end
    elseif opt == "i" or opt == "info" then
        print_info = true
    elseif (opt == "f") then
        fahrenheit = true
        print_temp = true
    elseif (opt == "t") then
        print_temp = true
    elseif opt == "V" or opt == "verbose" then
        verbose = tonumber(arg)
        likwid.setVerbosity(verbose)
    elseif (opt == "s") then
        time_interval = likwid.parse_time(arg) / 1000000
        time_orig = arg
        stethoscope = true
    elseif opt == "?" then
        print_stderr("Invalid commandline option -"..arg)
        os.exit(1)
    elseif opt == "!" then
        print_stderr("Option requires an argument")
        os.exit(1)
    end
end

-- initialize sysfeatures module
local err = likwid.initSysFeatures()
if err < 0 then
    print_stderr("Cannot initialize LIKWID sysfeatures")
    os.exit(1)
end

-- enumerate available devices
local device_types = { likwid.hwthread, likwid.core, likwid.die, likwid.socket, likwid.node }
if likwid.nvSupported() then
    table.insert(device_types, likwid.nvidia_gpu)
end
if likwid_rocmSupported() then
    table.insert(device_types, likwid.amd_gpu)
end

local devices = {}
for _, devtype in pairs(device_types) do
    -- getAvailableDevices returns a list of integers IDs for a particular device type
    devices[devtype] = {}
    for _, devid in pairs(likwid.getAvailableDevices(devtype)) do
        devices[devtype][devid] = likwid.createDevice(devtype, devid)
    end
end

-- set client access mode
if likwid.setAccessClientMode(access_mode) ~= 0 then
    os.exit(1)
end

-- check if stethoscope mode is enabled simultaneous with a program parameter
if #arg ~= 0 and stethoscope then
    print_stderr("Stethoscope mode cannot be used to launch programs")
    os.exit(1)
end

-- print device information
if print_info then
    print_stdout(likwid.hline);
    print_stdout(string.format("CPU name:\t%s", cpuinfo["osname"]))
    print_stdout(string.format("CPU type:\t%s", cpuinfo["name"]))
    if cpuinfo["clock"] > 0 then
        print_stdout(string.format("CPU clock:\t%3.2f GHz", cpuinfo["clock"] * 1.E-09))
    else
        print_stdout(string.format("CPU clock:\t%3.2f GHz", likwid.getCpuClock() * 1.E-09))
    end
    print_stdout(likwid.hline)
end

-- TODO print clock speeds and power information when print_info is enabled

-- Check if time interval is not too short TODO, we currently do not have a way to retrieve
-- the timeUnit and it may also differ between metric types.
if time_interval < 1 then
    print_stderr("Warning: Time interval too short, measurement may be inaccurate")
end

-- TODO begin: do we still need this?
local pinList = {}
local execString = ""
if use_perfctr then
    execString = string.format("<INSTALLED_PREFIX>/bin/likwid-perfctr -C %s -f -g CLOCK ",table.concat(pinList, "@"))
elseif not stethoscope then
    execString = string.format("<INSTALLED_PREFIX>/bin/likwid-pin -c %s -q ",table.concat(pinList, "@"))
end
-- TODO end

-- TODO begin: what does this do?
local execList = {}
if #arg == 0 then
    if use_perfctr then
        execString = execString .. string.format(" -S %s ", time_orig)
        stethoscope = false
    else
        stethoscope = true
    end
else
    for i=1, likwid.tablelength(arg)-2 do
        if string.find(arg[i], " ") then
            table.insert(execList, "\""..arg[i].."\"")
        else
            table.insert(execList, arg[i])
        end
    end
    execString = execString .. table.concat(execList," ")
end
-- TODO end

-- Read all metrics from power_metrics, but filter them by the features which are actually available
local power_features_wanted = {
    "rapl.pkg_energy", "rapl.dram_energy", "rapl.pp0_energy", "rapl.pp1_energy", "rapl.psys_energy",
    "nvml.energy",
}
local features_available = likwid.sysFeatures_list()
local power_features_available = {}
for _, wanted in pairs(power_features_wanted) do
    for _, available in pairs(features_available) do
        if wanted == available.FullName then
            table.insert(power_features_available, available)
            break
        end
    end
end

if #power_features_available == 0 then
    print_stderr("No power features available. Cannot measure power.")
    os.exit(1)
end

-- start measure time
local time_before = likwid.getTimeOfDay()
local prev_time = time_before
local energy = {} -- structure: energy[dev_type][dev_idx][metric_type][X] (X=1: raw measurement, X=2: corrected measurement)

next_power2 = function(x)
    local l2x = math.log(x) / math.log(2)
    return 2 ^ (math.ceil(l2x))
end

-- launch program (if any)
if not stethoscope then
    pid = likwid.startProgram(execString, 0, {})
    if not pid then
        print_stderr(string.format("Failed to execute %s!", execString))
        os.exit(1)
    end
    -- in order to avoid wait times when the program has exited, we cap the maximum wait time
    read_interval = math.min(read_interval, 1)
end

-- loop and measure until either the launched program exits or if the time elapses
local exitvalue = 0
local quit = false
while true do
    -- Iterate over all device types, devices, and features.
    for devtype, devlist in pairs(devices) do
        if energy[devtype] == nil then
            energy[devtype] = {}
        end
        energy_for_devtype = energy[devtype]
        for _, dev in pairs(devlist) do
            if energy_for_devtype[dev:id()] == nil then
                energy_for_devtype[dev:id()] = {}
            end
            energy_for_device = energy_for_devtype[dev:id()]
            for _, power_feature in pairs(power_features_available) do
                if power_feature.TypeID ~= devtype then
                    goto continue
                end
                if energy_for_device[power_feature.FullName] == nil then
                    energy_for_device[power_feature.FullName] = {}
                end
                energy_for_feature = energy_for_device[power_feature.FullName]
                -- Read old and new value.
                -- We have to keep track of raw and aggregated values separately in order to be able to notice
                -- if the counters overflow.
                local old_value_raw = energy_for_feature[1]
                local old_value_aggr = energy_for_feature[2]
                local new_value_raw, err = likwid.sysFeatures_get(power_feature.FullName, dev)
                new_value_raw = tonumber(new_value_raw)

                if new_value_raw ~= nil and old_value_aggr == nil then
                    -- if old value is not set, simply initialize
                    energy_for_feature[2] = 0
                elseif new_value_raw ~= nil and old_value_aggr ~= nil then
                    -- if old and new value are set, update energy, overflow adjusted
                    local delta = new_value_raw - old_value_raw
                    if new_value_raw < old_value_raw then
                        -- This case should only occur if the counter overflowed.
                        -- Get next power-of-two of old value in order to determine overflow point.
                        local overflow_point = next_power2(old_value)
                        delta = (new_value_raw + overflow_point) - old_value_raw
                    end
                    energy_for_feature[2] = energy_for_feature[2] + delta
                end

                -- unconditionally store the raw value, in order to do proper overflow detection
                energy_for_feature[1] = new_value_raw
                ::continue::
            end
        end
    end

    -- Instead of breaking out immediatley, we always delay breaking until here.
    -- This makes sure there is always a measurement taken after e.g. the program has exited.
    if quit then
        time_after = likwid.getTimeOfDay()
        break
    end

    -- Wait depending on the measurement mode
    if stethoscope then
        -- In stethoscope mode, we simply sleep periodically.
        local cur_time = likwid.getTimeOfDay()
        if cur_time > time_before + time_interval then
            quit = true
            sleep_duration = 0
        elseif cur_time - time_before + read_interval > time_interval then
            sleep_duration = time_interval - (cur_time - time_before)
        else
            sleep_duration = read_interval - (cur_time - prev_time)
        end
        sleep_duration = math.max(sleep_duration, 0)

        -- sleep accepts its parameters in microsecond range
        likwid.nanosleep(sleep_duration)
        prev_time = likwid.getTimeOfDay()
    else
        -- In execution wrapper mode, we simply check and wait
        if likwid.getSignalState() ~= 0 then
            likwid.killProgram()
            quit = true
        else
            exitvalue, exited = likwid.checkProgram(pid)
            if exited then
                io.stdout:flush()
                quit = true
            else
                likwid.sleep(read_interval)
            end
        end
    end
end

runtime = time_after - time_before

print_stdout(likwid.hline)
print_stdout(string.format("Runtime: %.2fs", runtime))

for dev_type, energy_for_devtype in pairs(energy) do
    -- First check if there actually are any features recorded for this device type.
    -- This avoids printing headers without any results.
    local has_measurement = false
    for _, energy_for_device in pairs(energy_for_devtype) do
        for _, energy_for_feature in pairs(energy_for_device) do
            if #energy_for_feature > 0 then
                has_measurement = true
                goto outer_break
            end
        end
    end
    ::outer_break::

    if not has_measurement then
        goto continue
    end

    -- At this point we know that devices[dev_type] contains at least one device.
    -- However, because dev_type is a string and doesn't always start at 0, we cannot simply index the first
    -- entry in devices[dev_type] to retrieve the device type name.
    -- To solve this, loop over devices[dev_type] and just remember the first entry.
    local first_dev_id = nil
    for dev_id, dev in pairs(devices[dev_type]) do
        first_dev_id = dev_id
        break
    end

    print_stdout(string.format("Device Type: %s", devices[dev_type][first_dev_id]:typeName()))
    for dev_id, energy_for_device in pairs(energy_for_devtype) do
        print_stdout(string.format("  ID: %s", dev_id))
        for feature_name, energy_for_feature in pairs(energy_for_device) do
            local joules = energy_for_feature[2]
            if energy_for_feature[2] == nil then
                joules = "ERR"
                watts = "ERR"
            else
                joules = string.format("%11.3f", joules)
                watts = string.format("%11.3f", joules / runtime)
            end

            print_stdout(string.format("    %s:", feature_name))
            print_stdout(string.format("      Energy consumed:    %s J", joules))
            print_stdout(string.format("      Average power draw: %s W", watts))
        end
    end
    ::continue::
end

print_stdout(likwid.hline)

if print_temp then
    print_stderr("Printing temperature is not yet implemented")
    os.exit(1)
end

likwid.finalizeSysFeatures()
likwid.finalize()
os.exit(exitvalue)

File: src/applications/likwid-powermeter.c
/*
 * =======================================================================================
 *
 *      Filename:  likwid-powermeter.c
 *
 *      Description:  An application to get information about power 
 *      consumption on architectures implementing the RAPL interface.
 *
 *      Version:   <VERSION>
 *      Released:  <DATE>
 *
 *      Author:  Jan Treibig (jt), jan.treibig@gmail.com
 *      Project:  likwid
 *
 *      Copyright (C) 2014 Jan Treibig
 *
 *      This program is free software: you can redistribute it and/or modify it under
 *      the terms of the GNU General Public License as published by the Free Software
 *      Foundation, either version 3 of the License, or (at your option) any later
 *      version.
 *
 *      This program is distributed in the hope that it will be useful, but WITHOUT ANY
 *      WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 *      PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 *      You should have received a copy of the GNU General Public License along with
 *      this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * =======================================================================================
 */
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <sched.h>
#include <sys/types.h>
#include <unistd.h>
#include <ctype.h>

#include <types.h>
#include <strUtil.h>
#include <error.h>
#include <lock.h>
#include <timer.h>
#include <cpuid.h>
#include <numa.h>
#include <accessClient.h>
#include <msr.h>
#include <affinity.h>
#include <perfmon.h>
#include <power.h>
#include <thermal.h>

/* #####   MACROS  -  LOCAL TO THIS SOURCE FILE   ######################### */

#define HELP_MSG \
fprintf(stdout, "\nlikwid-powermeter --  Version  %d.%d \n\n",VERSION,RELEASE); \
fprintf(stdout, "A tool to print Power and Clocking information on Intel SandyBridge CPUS.\n"); \
fprintf(stdout, "Options:\n"); \
fprintf(stdout, "-h\t\t Help message\n"); \
fprintf(stdout, "-v\t\t Version information\n"); \
fprintf(stdout, "-M <0|1>\t set how MSR registers are accessed: 0=direct, 1=msrd \n"); \
fprintf(stdout, "-c <list>\t specify sockets to measure\n"); \
fprintf(stdout, "-i\t\t print information from MSR_PKG_POWER_INFO register and Turbo Mode\n"); \
fprintf(stdout, "-s <duration>\t set measure duration in sec. (default 2s) \n"); \
fprintf(stdout, "-p\t\t print dynamic clocking and CPI values (requires executable)\n\n");   \
fprintf(stdout, "Usage: likwid-powermeter -s 4 -c 1 \n");  \
fprintf(stdout, "Alternative as wrapper: likwid-powermeter -c 1 ./a.out\n"); \
fflush(stdout);

#define VERSION_MSG \
fprintf(stdout, "likwid-powermeter  %d.%d \n\n",VERSION,RELEASE); \
fflush(stdout);


int main (int argc, char** argv)
{
    int socket_fd = -1;
    int optInfo = 0;
    int optClock = 0;
    int optStethoscope = 0;
    int optSockets = 0;
    double runtime;
    int hasDRAM = 0;
    int hasPP0 = 0;
    int c;
    bstring argString;
    bstring eventString = bfromcstr("CLOCK");
    int numSockets=1;
    int numThreads=0;
    int threadsSockets[MAX_NUM_NODES*2];
    int threads[MAX_NUM_THREADS];
    threadsSockets[0] = 0;

    if (argc == 1)
    {
    	HELP_MSG;
    	exit (EXIT_SUCCESS);
    }

    while ((c = getopt (argc, argv, "+c:hiM:ps:v")) != -1)
    {
        switch (c)
        {
            case 'c':
                CHECK_OPTION_STRING;
                numSockets = bstr_to_cpuset_physical((uint32_t*) threadsSockets, argString);
                bdestroy(argString);
                optSockets = 1;
                break;

            case 'h':
                HELP_MSG;
                exit (EXIT_SUCCESS);
            case 'i':
                optInfo = 1;
                break;
            case 'M':  /* Set MSR Access mode */
                CHECK_OPTION_STRING;
                accessClient_setaccessmode(str2int((char*) argString->data));
                bdestroy(argString);
                break;
            case 'p':
                optClock = 1;
                break;
            case 's':
                CHECK_OPTION_STRING;
                optStethoscope = str2int((char*) argString->data);
                bdestroy(argString);
                break;
            case 'v':
                VERSION_MSG;
                exit (EXIT_SUCCESS);
            case '?':
            	if (optopt == 's' || optopt == 'M' || optopt == 'c')
            	{
            		HELP_MSG;
            	}
                else if (isprint (optopt))
                {
                    fprintf (stderr, "Unknown option `-%c'.\n", optopt);
                }
                else
                {
                    fprintf (stderr,
                            "Unknown option character `\\x%x'.\n",
                            optopt);
                }
                exit( EXIT_FAILURE);
            default:
                HELP_MSG;
                exit (EXIT_SUCCESS);
        }
    }

    if (!lock_check())
    {
        fprintf(stderr,"Access to performance counters is locked.\n");
        exit(EXIT_FAILURE);
    }
    if (optClock && optind == argc)
    {
    	fprintf(stderr,"Commandline option -p requires an executable.\n");
    	exit(EXIT_FAILURE);
    }
    if (optSockets && !optStethoscope && optind == argc)
    {
        fprintf(stderr,"Commandline option -c requires an executable if not used in combination with -s.\n");
        exit(EXIT_FAILURE);
    }
    if (optStethoscope == 0 && optind == argc && !optInfo)
    {
        fprintf(stderr,"Either -s <seconds> or executable must be given on commandline.\n");
        exit(EXIT_FAILURE);
    }

    if (cpuid_init() == EXIT_FAILURE)
    {
        fprintf(stderr, "CPU not supported\n");
        exit(EXIT_FAILURE);
    }
    
    if (numSockets > cpuid_topology.numSockets)
    {
    	fprintf(stderr, "System has only %d sockets but %d are given on commandline.\n",
    			cpuid_topology.numSockets, numSockets);
    	exit(EXIT_FAILURE);
    }

    numa_init(); /* consider NUMA node as power unit for the moment */
    accessClient_init(&socket_fd);
    msr_init(socket_fd);
    timer_init();

    /* check for supported processors */
    if ((cpuid_info.model == SANDYBRIDGE_EP) ||
            (cpuid_info.model == SANDYBRIDGE) ||
            (cpuid_info.model == IVYBRIDGE) ||
            (cpuid_info.model == IVYBRIDGE_EP) ||
            (cpuid_info.model == HASWELL) ||
            (cpuid_info.model == HASWELL_EX) ||
            (cpuid_info.model == NEHALEM_BLOOMFIELD) ||
            (cpuid_info.model == NEHALEM_LYNNFIELD) ||
            (cpuid_info.model == NEHALEM_WESTMERE) ||
            (cpuid_info.model == ATOM_SILVERMONT))
    {
        if (numSockets == 0)
        {
            numSockets = numa_info.numberOfNodes;
        }
        for(int i=0; i<numSockets; i++)
        {
            power_init(numa_info.nodes[threadsSockets[i]].processors[0]);
        }
    }
    else
    {
        fprintf (stderr, "Query Turbo Mode only supported on Intel Nehalem/Westmere/SandyBridge/IvyBridge/Haswell processors!\n");
        exit(EXIT_FAILURE);
    }

    double clock = (double) timer_getCpuClock();

    fprintf(stdout, HLINE);
    fprintf(stdout, "CPU name:\t%s \n",cpuid_info.name);
    fprintf(stdout, "CPU clock:\t%3.2f GHz \n",  (float) clock * 1.E-09);
    fprintf(stdout, HLINE);
    fflush(stdout);

    if (optInfo)
    {
        if (power_info.turbo.numSteps != 0)
        {
            fprintf(stdout, "Base clock:\t%.2f MHz \n",  power_info.baseFrequency );
            fprintf(stdout, "Minimal clock:\t%.2f MHz \n",  power_info.minFrequency );
            fprintf(stdout, "Turbo Boost Steps:\n");
            for (int i=0; i < power_info.turbo.numSteps; i++ )
            {
                fprintf(stdout, "C%d %.2f MHz \n",i+1,  power_info.turbo.steps[i] );
            }
        }
        fprintf(stdout, HLINE);
        fflush(stdout);
    }

    if ((cpuid_info.model == SANDYBRIDGE_EP) ||
        (cpuid_info.model == IVYBRIDGE_EP) ||
        (cpuid_info.model == HASWELL_EX))
    {
        hasDRAM = 1;
    }
    if (cpuid_info.model != ATOM_SILVERMONT)
    {
        hasPP0 = 1;
    }
    if ((cpuid_info.model != SANDYBRIDGE) &&
        (cpuid_info.model != SANDYBRIDGE_EP)  &&
        (cpuid_info.model != IVYBRIDGE)  &&
        (cpuid_info.model != IVYBRIDGE_EP)  &&
        (cpuid_info.model != HASWELL) &&
        (cpuid_info.model != HASWELL_EX) &&
        (cpuid_info.model != ATOM_SILVERMONT))
    {
        fprintf (stderr, "RAPL not supported on this processor!\n");
        exit(EXIT_FAILURE);
    }

    if (optInfo)
    {
        fprintf(stdout, "Thermal Spec Power: %g Watts \n", power_info.tdp );
        fprintf(stdout, "Minimum  Power: %g Watts \n", power_info.minPower);
        fprintf(stdout, "Maximum  Power: %g Watts \n", power_info.maxPower);
        fprintf(stdout, "Maximum  Time Window: %g micro sec \n", power_info.maxTimeWindow);
        fprintf(stdout, HLINE);
        fflush(stdout);
        exit(EXIT_SUCCESS);
    }

    if (optClock)
    {
        affinity_init();
        argString = bformat("S%u:0-%u", threadsSockets[0], cpuid_topology.numCoresPerSocket-1);
        for (int i=1; i<numSockets; i++)
        {
            bstring tExpr = bformat("@S%u:0-%u", threadsSockets[i], cpuid_topology.numCoresPerSocket-1);
            bconcat(argString, tExpr);
        }
        numThreads = bstr_to_cpuset(threads, argString);
        bdestroy(argString);
        perfmon_init(numThreads, threads, stdout);
        perfmon_setupEventSet(eventString, NULL);
    }

    {
        PowerData pDataPkg[MAX_NUM_NODES*2];
        PowerData pDataDram[MAX_NUM_NODES*2];
        PowerData pDataPP0[MAX_NUM_NODES*2];
        fprintf(stdout, "Measure on sockets: %d", threadsSockets[0]);
        for (int i=1; i<numSockets; i++)
        {
            fprintf(stdout, ", %d", threadsSockets[i]);
        }
        fprintf(stdout, "\n");
        fflush(stdout);

        if (optStethoscope)
        {
            if (optClock)
            {
                perfmon_startCounters();
            }
            else
            {
                for (int i=0; i<numSockets; i++)
                {
                    int cpuId = numa_info.nodes[threadsSockets[i]].processors[0];
                    if (hasDRAM) power_start(pDataDram+i, cpuId, DRAM);
                    if (hasPP0) power_start(pDataPP0+i, cpuId, PP0);
                    power_start(pDataPkg+i, cpuId, PKG);
                }
            }
            sleep(optStethoscope);

            if (optClock)
            {
                perfmon_stopCounters();
                perfmon_printCounterResults();
                perfmon_finalize();
            }
            else
            {
                for (int i=0; i<numSockets; i++)
                {
                    int cpuId = numa_info.nodes[threadsSockets[i]].processors[0];
                    power_stop(pDataPkg+i, cpuId, PKG);
                    if (hasPP0) power_stop(pDataPP0+i, cpuId, PP0);
                    if (hasDRAM) power_stop(pDataDram+i, cpuId, DRAM);
                }
            }
            runtime = (double) optStethoscope;
        }
        else
        {
            TimerData time;
            argv +=  optind;
            bstring exeString = bfromcstr(argv[0]);

            for (int i=1; i<(argc-optind); i++)
            {
                bconchar(exeString, ' ');
                bcatcstr(exeString, argv[i]);
            }
            fprintf(stdout, "Executing: %s\n",bdata(exeString));
            fflush(stdout);


            if (optClock)
            {
                perfmon_startCounters();
            }
            else
            {
                for (int i=0; i<numSockets; i++)
                {
                    int cpuId = numa_info.nodes[threadsSockets[i]].processors[0];
                    if (hasDRAM) power_start(pDataDram+i, cpuId, DRAM);
                    if (hasPP0) power_start(pDataPP0+i, cpuId, PP0);
                    power_start(pDataPkg+i, cpuId, PKG);
                }

                timer_start(&time);
            }

            if (system(bdata(exeString)) == EOF)
            {
                fprintf(stderr, "Failed to execute %s!\n", bdata(exeString));
                exit(EXIT_FAILURE);
            }

            if (optClock)
            {
                perfmon_stopCounters();
                perfmon_printCounterResults();
                perfmon_finalize();
            }
            else
            {
                timer_stop(&time);

                for (int i=0; i<numSockets; i++)
                {
                    int cpuId = numa_info.nodes[threadsSockets[i]].processors[0];
                    power_stop(pDataPkg+i, cpuId, PKG);
                    if (hasDRAM) power_stop(pDataDram+i, cpuId, DRAM);
                    if (hasPP0) power_stop(pDataPP0+i, cpuId, PP0);
                }
                runtime = timer_print(&time);
            }
        }

        if (!optClock)
        {
            fprintf(stdout, "Runtime: %g second \n",runtime);
            fprintf(stdout, HLINE);
            for (int i=0; i<numSockets; i++)
            {
                fprintf(stdout, "Socket %d\n",threadsSockets[i]);
                fprintf(stdout, "Domain: PKG \n");
                fprintf(stdout, "Energy consumed: %g Joules \n", power_printEnergy(pDataPkg+i));
                fprintf(stdout, "Power consumed: %g Watts \n", power_printEnergy(pDataPkg+i) / runtime );
                if (hasDRAM)
                {
                    fprintf(stdout, "Domain: DRAM \n");
                    fprintf(stdout, "Energy consumed: %g Joules \n", power_printEnergy(pDataDram+i));
                    fprintf(stdout, "Power consumed: %g Watts \n", power_printEnergy(pDataDram+i) / runtime );
                }
                if (hasPP0)
                {
                    fprintf(stdout, "Domain: PP0 \n");
                    fprintf(stdout, "Energy consumed: %g Joules \n", power_printEnergy(pDataPP0+i));
                    fprintf(stdout, "Power consumed: %g Watts \n", power_printEnergy(pDataPP0+i) / runtime );
                }
                fprintf(stdout, "\n");
            }
            fflush(stdout);
        }
    }

#if 0
    if ( cpuid_hasFeature(TM2) )
    {
        thermal_init(0);
        printf("Current core temperatures:\n");

        for (uint32_t i = 0; i < cpuid_topology.numCoresPerSocket; i++ )
        {
            printf("Core %d: %u C\n",
                    numa_info.nodes[socketId].processors[i],
                    thermal_read(numa_info.nodes[socketId].processors[i]));
        }
    }
#endif

    msr_finalize();
    return EXIT_SUCCESS;
}

